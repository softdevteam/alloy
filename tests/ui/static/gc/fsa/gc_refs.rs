#![feature(gc)]
#![feature(negative_impls)]
#![allow(dead_code)]
#![allow(unused_variables)]
include!{"./auxiliary/types.rs"}

impl Drop for HasGc {
    fn drop(&mut self) {
        use_val(self.a); // should fail
        use_val(self.b); // should pass
        use_val(self.c[0]); // should fail

        let a = self.a; // should fail
        let b = self.b; // should pass
        let c = self.c;
        use_val(c[1]); // should fail

        // should pass, as not a field projection
        let c = Gc::new(1);
        use_val(c);
    }
}

fn main() {
    Gc::new(HasGc::default());
    //~^     ERROR: The drop method for `HasGc` cannot be safely finalized.
    //~^^    ERROR: The drop method for `HasGc` cannot be safely finalized.
    //~^^^   ERROR: The drop method for `HasGc` cannot be safely finalized.
    //~^^^^  ERROR: The drop method for `HasGc` cannot be safely finalized.
    //~^^^^^ ERROR: The drop method for `HasGc` cannot be safely finalized.
}
