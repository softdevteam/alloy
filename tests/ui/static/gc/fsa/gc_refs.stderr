error: The drop method for `HasGc` cannot be safely finalized.
  --> $DIR/gc_refs.rs:21:13
   |
LL |         use_val(self.a); // should fail
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `Gc<u64>`
   |                 from a drop method because it might have already been finalized.
...
LL |     Gc::new(HasGc::default());
   |     --------^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasGc>` here.

error: The drop method for `HasGc` cannot be safely finalized.
  --> $DIR/gc_refs.rs:21:13
   |
LL |         use_val(self.c[0]); // should fail
   |                 ---------
   |                 |
   |                 a finalizer cannot safely dereference this `Gc<u64>`
   |                 from a drop method because it might have already been finalized.
...
LL |     Gc::new(HasGc::default());
   |     --------^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasGc>` here.

error: The drop method for `HasGc` cannot be safely finalized.
  --> $DIR/gc_refs.rs:21:13
   |
LL |         let a = self.a; // should fail
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `Gc<u64>`
   |                 from a drop method because it might have already been finalized.
...
LL |     Gc::new(HasGc::default());
   |     --------^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasGc>` here.

error: The drop method for `HasGc` cannot be safely finalized.
  --> $DIR/gc_refs.rs:21:13
   |
LL |         use_val(c[1]); // should fail
   |                 ----
   |                 |
   |                 a finalizer cannot safely dereference this `Gc<u64>`
   |                 from a drop method because it might have already been finalized.
...
LL |     Gc::new(HasGc::default());
   |     --------^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasGc>` here.

error: aborting due to 4 previous errors

